spec/lib/diaspora_federation/discovery/discovery_spec.rb

```diff
@@ -88,66 +91,148 @@ module DiasporaFederation
           callback_person = person
         end
 
-        expect(Discovery::Discovery.new(account).fetch_and_save).to be(callback_person)
+        expect(subject.fetch_and_save).to be(callback_person)
       end
 
       it "falls back to http if https fails with 404" do
```

Comment:
Does it make sense to fall back to HTTP in case if the same URI at HTTPS replied as `404`? We expect to have the same resource for HTTP and HTTPS at the same address, so if HTTPS is `404` and HTTP is not then it's likely to be some kind of evil :imp:.

Also, RFC 7033 says
> A WebFinger request is an HTTPS request to a WebFinger resource.  A
>   WebFinger resource is a well-known URI [3] using the HTTPS scheme
>   constructed along with the required query target and optional link
>   relation types.  WebFinger resources MUST NOT be served with any
>   other URI scheme (such as HTTP).

which we clearly violate this way. Is there any special reason why we need the HTTP fallback at all?

Reply:
We always did it that way. We allow pods to run on http (while we strongly recommend https!), so we need to allow to fetch webfinger via http. The 404 case maybe doesn't make much sense, but there are webservers configured to only work with http and 404 everything on https (I saw a pod like this once). We maybe need to discuss if we want to block such misconfigurations (and enforce https everywhere), but currently we allow that (and did always in the past).

Reply:
What we did in this case was fallback on XML/XRD for non-SSL sites. That way everything still works and it's spec-compliant, although it does mean maintaining the XRD service.  I notice a couple of projects that generate both the json webfinger and the personal xrd from a common backend structure which would relieve a lot of the maintenance problem - if you see that as a problem.

Reply:
> We always did it that way. 

But we didn't have RFC 7033 support before and RFC 7033 support requires to use HTTPS. So in order to be spec compiant we shouldn't provide this WebFinger route when pod doesn't have HTTPS available, and therefore shouldn't fetch it via HTTP.

You introduce RFC 7033 in this PR that's why it is in scope of this PR.

So I think the right thing to do is to support only the old draft for HTTP pods. It explicitly allows this:
>   WebFinger itself is agnostic regarding the scheme of such a URI: it
>   could be an "acct" URI as defined in the next section, an "http" or
>   "https" URI, a "mailto" URI, or some other scheme.

If we ever want to drop the old WebFinger, then we'll have to consider dropping HTTP support at all.

Also the draft talks about signatures for HTTP:
>   When using HTTP to request an XRD document, WebFinger clients SHOULD
>   verify the XRD document's signature, if present, to ensure that the
>   XRD document has not been modified.  Additionally, WebFinger servers
>   SHOULD include a signature for XRD documents served over HTTP.

But I can't find the signing procedure described, mainly, how to perform key exchange, which doesn't make much sense to do over the same HTTP channel (that's why it is probably not in the final spec).

What do you think?

Reply:
If we drop the "old WebFinger", we have to do it in a major release. So it makes sense to me to drop HTTP support in the same time. (Something we should probably do, now that browsers dislike http a lot and let's encrypt make damn easy to set up https).

So I would go for "https only for the "new" webfinger, if https is not available then fallback to the old one".

Reply:
If we want to remove http support we can add the RFC 7033 WebFinger without http support, and fallback to the old WebFinger if needed. But then we should drop http support at the next major (so drop old WebFinger support with it, at least stop fetching it), otherwise we win absolutely no security if it is still possible to downgrade to the old way without https. But then we should start warning podmins now, so they have time to add https if they don't have it already.

What should we do for development? I think we should at least allow http fallback for development, because we don't have https there. Keeping the old WebFinger for development but using the RFC 7033 WebFinger in production doesn't sound like a good idea to me.

Reply:
Maybe we should make the http fallback for RFC 7033 configurable and off by default? So by default we don't fall back to HTTP for the new WebFinger and if the `diaspora_federation` gem based application opts in, then we do fall back. We can even set it to fallback in diaspora itself (override the default), but for `diaspora_federation` it will be off by default, so our library is RFC 7033 compliant by default with no exceptions. For diaspora we could allow this fallback either for both production and development, or for development only, or we can have it on until we support http and turn it off at the time we drop http whenever this happens and leave it just for development purposes.

Reply:
That is how it works for OpenID Connect gems I worked with: you can use it with HTTP, until you explicitly set some magic setting that tells the gem to ignore https only restriction.

Reply:
OK, so I'll add a setting for the fallback, and then we can handle that in diaspora.

Reply:
Done and PR updated.